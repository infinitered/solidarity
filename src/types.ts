import { GluegunRunContext } from 'gluegun'
export const solidarity = {
  binaryExists: require('./extensions/functions/binaryExists'),
  getSolidaritySettings: require('./extensions/functions/getSolidaritySettings'),
  setSolidaritySettings: require('./extensions/functions/setSolidaritySettings'),
  checkRequirement: require('./extensions/functions/checkRequirement'),
  updateRequirement: require('./extensions/functions/updateRequirement'),
  skipRule: require('./extensions/functions/skipRule'),
  checkENV: require('./extensions/functions/checkENV'),
  checkCLI: require('./extensions/functions/checkCLI'),
  checkDir: require('./extensions/functions/checkDir'),
  checkFile: require('./extensions/functions/checkFile'),
  updateVersions: require('./extensions/functions/updateVersions'),
  removeNonVersionCharacters: require('./extensions/functions/removeNonVersionCharacters'),
  getVersion: require('./extensions/functions/getVersion'),
  getLineWithVersion: require('./extensions/functions/getLineWithVersion')
}

export interface SolidarityPlugin {
  name: string
  description: string
  snapshot: string | SnapshotType
}

export interface SolidarityRunContext extends GluegunRunContext {
  solidarity: typeof solidarity
  _pluginsList: Array<SolidarityPlugin & {templateDirectory: string}>
  addPlugin: (config: SolidarityPlugin) => void
  printSeparator: () => void
  outputMode: SolidarityOutputMode
}

export type SnapshotType = (context: SolidarityRunContext) => Promise<void>

export interface SolidarityRule {
  rule: 'cli' | 'env' | 'dir' | 'file'
  binary?: string
  semver?: string
  version?: string
  line?: string | number
  variable?: string
  location?: string
  error?: string
  platform?: string | string[]
  matchIndex?: number
}

export type SolidarityRequirement = SolidarityRule[]

export interface SolidaritySettings {
  requirements: object
  config: object
}

export enum SolidarityOutputMode {
  MODERATE,
  VERBOSE,
  SILENT
}
