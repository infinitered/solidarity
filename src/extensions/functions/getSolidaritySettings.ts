import { SolidarityRunContext, SolidaritySettings } from '../../types'
module.exports = (context: SolidarityRunContext): SolidaritySettings => {
  const { filesystem } = context

  // for now only JSON support
  let solidaritySettings
  if (filesystem.exists('.solidarity')) {
    solidaritySettings = JSON.parse(filesystem.read('.solidarity'))
  } else if (filesystem.exists('.solidarity.json')) {
    solidaritySettings = JSON.parse(filesystem.read('.solidarity.json'))
  } else {
    // if we got here there was no solidarity file
    throw 'ERROR: No solidarity file was found'
  }

  // Check shape
  if (solidaritySettings.requirements) {
    return solidaritySettings
  } else {
    throw 'ERROR: Found, but no requirements key.  Please validate your solidarity file'
  }
}
