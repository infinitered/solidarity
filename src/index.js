const { build } = require('gluegun')
const {dependenciesMet} = require('gluegun/sniff')
if(dependenciesMet==1){
    module.exports = async () => {
      // setup the runtime
      build()
        .brand('solidarity')
        .src(__dirname)
        // It might also be worth checking global
        // `npm config get prefix` returns '/usr/local'
        // global npm is stored /usr/local/lib/node_modules
        .plugins('./node_modules', { matching: 'solidarity-*', hidden: true })
        // for testing - force load a plugin
        // .plugin('../solidarity-react-native')
        .create()
        .run()
    }
}
