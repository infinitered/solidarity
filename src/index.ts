import { build } from 'gluegun'

module.exports = async (): Promise<void> => {
  // setup the runtime
  build()
    .brand('solidarity')
    .src(__dirname)
    // local installs
    .plugins('./node_modules', { matching: 'solidarity-*', hidden: true })
    // global installs
    .plugins('/usr/local/lib/node_modules', { matching: 'solidarity-*', hidden: true })
    // for testing - force load a local plugin
    // .plugin('../solidarity-react-native')
    .create()
    .run()
}

export * from './types'
