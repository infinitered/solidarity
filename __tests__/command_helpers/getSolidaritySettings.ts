import { solidarity } from '../../src';
import getSolidaritySettings from '../../src/extensions/functions/getSolidaritySettings'

const context = require('gluegun')

describe('getSolidaritySettings', () => {
  describe('w/ .solidarity', () => {
    test('getSolidaritySettings exists', () => expect(getSolidaritySettings).toMatchSnapshot())
    
    test('getSolidaritySettings succeeds', async () => {
      const resultSettings = getSolidaritySettings(context)
      // we got an object with requirements defined
      expect(resultSettings).toMatchObject({requirements: {}})
    })
  })
  
  test('getSolidaritySettings can fail', async () => {
    expect(() => {
      process.chdir('__tests__')
      const resultSettings = getSolidaritySettings(context)
    }).toThrow()
    process.chdir('../')
  })
})

