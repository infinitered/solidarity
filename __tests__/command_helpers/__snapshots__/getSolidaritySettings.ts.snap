// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSolidaritySettings w/ .solidarity getSolidaritySettings exists 1`] = `[Function]`;
